fn.go

```diff
@@ -229,12 +229,46 @@ func ComposedFromYAML(y string) (map[string]*fnv1.Resource, error) {
 	return out, nil
 }
 
+// stripMarkdownCodeBlocks removes markdown code block syntax from the input string.
+// LLMs sometimes wrap JSON/YAML output in markdown code blocks which breaks parsing.
+func stripMarkdownCodeBlocks(s string) string {
+	s = strings.TrimSpace(s)
+	s = strings.TrimPrefix(s, "```json")
+	s = strings.TrimPrefix(s, "```yaml")
+	s = strings.TrimPrefix(s, "```")
+	s = strings.TrimSuffix(s, "```")
+	return strings.TrimSpace(s)
+}
+
+// extractJSONFromAgentError attempts to extract JSON from langchaingo agent parsing errors.
+// When the agent framework fails to parse output, it returns an error containing the raw output.
+// This function extracts and cleans that output for a second parsing attempt.
+// Returns the extracted content and true if extraction succeeded, empty string and false otherwise.
+func extractJSONFromAgentError(err error) (string, bool) {
+	if err == nil {
+		return "", false
+	}
+	if !strings.Contains(err.Error(), "unable to parse agent output") {
+		return "", false
+	}
+	parts := strings.SplitN(err.Error(), "unable to parse agent output: ", 2)
+	if len(parts) != 2 {
+		return "", false
+	}
+	cleaned := stripMarkdownCodeBlocks(parts[1])
+	return cleaned, true
+}
```

Comment:
Is there any possible path where we don't have to do string parsing? e.g. are the errors structured in any way?